File: clients/cpp/src/api_client.cpp
#include "orbit/api_client.hpp"
#include <curl/curl.h>
#include <sstream>
#include <stdexcept>

namespace orbit {

ApiClient::ApiClient(std::string api_url, std::string api_key, std::string session_id)
  : api_url_(std::move(api_url)), api_key_(std::move(api_key)), session_id_(std::move(session_id)) {}

std::string ApiClient::endpoint() const {
  if (api_url_.size() >= 8 && api_url_.rfind("/v1/chat") == api_url_.size() - 8) return api_url_;
  if (!api_url_.empty() && api_url_.back() == '/') return api_url_ + "v1/chat";
  return api_url_ + "/v1/chat";
}

struct WriteCtx {
  std::function<void(const StreamResponse&)> cb;
  std::string buffer;
};

static size_t on_write(char* ptr, size_t size, size_t nmemb, void* userdata) {
  auto* ctx = static_cast<WriteCtx*>(userdata);
  size_t total = size * nmemb;
  ctx->buffer.append(ptr, total);
  size_t pos = 0;
  for (;;) {
    auto nl = ctx->buffer.find('\n', pos);
    if (nl == std::string::npos) break;
    std::string line = ctx->buffer.substr(pos, nl - pos);
    pos = nl + 1;
    if (line.rfind("data: ", 0) == 0) {
      std::string payload = line.substr(6);
      if (payload == "[DONE]" || payload.empty()) {
        ctx->cb(StreamResponse{"", true});
        continue;
      }
      // naive extraction of response text; for robust use a JSON parser
      auto key = std::string{"\"response\":"};
      auto kpos = payload.find(key);
      if (kpos != std::string::npos) {
        auto start = payload.find('"', kpos + key.size());
        if (start != std::string::npos) {
          auto end = payload.find('"', start + 1);
          if (end != std::string::npos) {
            ctx->cb(StreamResponse{payload.substr(start + 1, end - start - 1), false});
          }
        }
      } else {
        ctx->cb(StreamResponse{payload, false});
      }
    }
  }
  // keep remainder
  ctx->buffer = ctx->buffer.substr(pos);
  return total;
}

void ApiClient::stream_chat(const std::string& message, bool stream, std::function<void(const StreamResponse&)> on_chunk) {
  CURL* curl = curl_easy_init();
  if (!curl) throw std::runtime_error("Failed to init curl");

  std::string url = endpoint();
  curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
  curl_easy_setopt(curl, CURLOPT_POST, 1L);

  std::ostringstream os;
  os << "{\"messages\":[{\"role\":\"user\",\"content\":";
  // very naive escape of quotes/newlines
  for (char c : message) {
    if (c == '\\') os << "\\\\";
    else if (c == '"') os << "\\\"";
    else if (c == '\n') os << "\\n";
    else os << c;
  }
  os << "}],\"stream\":" << (stream ? "true" : "false") << "}";
  auto payload = os.str();
  curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());

  struct curl_slist* headers = nullptr;
  headers = curl_slist_append(headers, "Content-Type: application/json");
  headers = curl_slist_append(headers, stream ? "Accept: text/event-stream" : "Accept: application/json");
  if (!api_key_.empty()) {
    headers = curl_slist_append(headers, ("X-API-Key: " + api_key_).c_str());
  }
  if (!session_id_.empty()) {
    headers = curl_slist_append(headers, ("X-Session-ID: " + session_id_).c_str());
  }
  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

  WriteCtx ctx{on_chunk, {}};
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, on_write);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &ctx);

  CURLcode res = curl_easy_perform(curl);
  curl_slist_free_all(headers);
  curl_easy_cleanup(curl);
  if (res != CURLE_OK) {
    throw std::runtime_error("Request failed");
  }
}

} // namespace orbit

File: clients/php/src/ApiClient.php
<?php

namespace SchmiTech\Orbit;

class StreamResponse {
    public string $text; public bool $done;
    public function __construct(string $text, bool $done) { $this->text = $text; $this->done = $done; }
}

class ApiClient {
    private string $apiUrl; private ?string $apiKey; private ?string $sessionId;
    public function __construct(string $apiUrl, ?string $apiKey = null, ?string $sessionId = null) {
        $this->apiUrl = $apiUrl; $this->apiKey = $apiKey; $this->sessionId = $sessionId;
    }
    private function endpoint(): string { return str_ends_with($this->apiUrl, '/v1/chat') ? $this->apiUrl : rtrim($this->apiUrl, '/') . '/v1/chat'; }
    public function setSessionId(?string $id): void { $this->sessionId = $id; }

    public function streamChat(string $message, bool $stream, callable $onChunk): void {
        $ch = curl_init($this->endpoint());
        $headers = [
            'Content-Type: application/json',
            'Accept: ' . ($stream ? 'text/event-stream' : 'application/json'),
            'X-Request-ID: ' . (string) (int) (microtime(true)*1000)
        ];
        if ($this->apiKey) $headers[] = 'X-API-Key: ' . $this->apiKey;
        if ($this->sessionId) $headers[] = 'X-Session-ID: ' . $this->sessionId;
        $body = json_encode(['messages' => [['role' => 'user', 'content' => $message]], 'stream' => $stream]);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_POSTFIELDS => $body,
            CURLOPT_RETURNTRANSFER => false,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_WRITEFUNCTION => function($ch, $data) use ($onChunk) {
                static $buffer = '';
                $buffer .= $data;
                while (($pos = strpos($buffer, "\n")) !== false) {
                    $line = trim(substr($buffer, 0, $pos));
                    $buffer = substr($buffer, $pos + 1);
                    if ($line === '') continue;
                    if (str_starts_with($line, 'data: ')) {
                        $json = trim(substr($line, 6));
                        if ($json === '' || $json === '[DONE]') { $onChunk(new StreamResponse('', true)); continue; }
                        $obj = json_decode($json, true);
                        if (is_array($obj) && isset($obj['response'])) {
                            $done = isset($obj['done']) ? (bool)$obj['done'] : false;
                            $onChunk(new StreamResponse((string)$obj['response'], $done));
                            if ($done) $onChunk(new StreamResponse('', true));
                        } else {
                            $onChunk(new StreamResponse($json, false));
                        }
                    } else {
                        $onChunk(new StreamResponse($line, false));
                    }
                }
                return strlen($data);
            }
        ]);
        $ok = curl_exec($ch);
        if ($ok === false) throw new \RuntimeException('cURL error: ' . curl_error($ch));
        curl_close($ch);
    }
}

